app/src/main/java/com/google/android/sambadocumentsprovider/browsing/broadcast/BroadcastUtils.java

```diff
@@ -84,55 +91,65 @@
-    ByteBuffer buffer = ByteBuffer.wrap(data);
+  static List<String> extractServers(byte[] data, int expectedTransId) throws BrowsingException {
+    List<String> servers = new ArrayList<>();
 
-    skipBytes(buffer, 2); // Skip transId.
-    skipBytes(buffer, 2); // Skip flags.
+    try {
+      ByteBuffer buffer = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN);
 
-    skipBytes(buffer, 2); // No questions.
-    skipBytes(buffer, 2); // Skip answers count.
-    skipBytes(buffer, 2); // No authority resources.
-    skipBytes(buffer, 2); // No additional resources.
+      int transId = buffer.getChar();
+      if (transId != expectedTransId) {
+        // This response is not to our broadcast.
 
-    int nameLength = buffer.get();
-    skipBytes(buffer, nameLength);
+        if (BuildConfig.DEBUG) Log.d(TAG, "Irrelevant broadcast response");
 
-    skipBytes(buffer, 1);
+        return servers;
+      }
 
-    int nodeStatus = buffer.getChar();
-    if (nodeStatus != 0x20 && nodeStatus != 0x21) {
-      throw new BrowsingException("Received negative response for the broadcast");
-    }
+      skipBytes(buffer, 2); // Skip flags.
 
-    skipBytes(buffer, 2);
-    skipBytes(buffer, 4);
-    skipBytes(buffer, 2);
+      skipBytes(buffer, 2); // No questions.
+      skipBytes(buffer, 2); // Skip answers count.
+      skipBytes(buffer, 2); // No authority resources.
+      skipBytes(buffer, 2); // No additional resources.
 
-    int addressListEntryCount = buffer.get();
+      int nameLength = buffer.get();
+      skipBytes(buffer, nameLength);
 
-    List<String> servers = new ArrayList<>();
+      skipBytes(buffer, 1);
 
-    for (int i = 0; i < addressListEntryCount; i++) {
-      StringBuilder nameBuilder = new StringBuilder();
-      for (int j = 0; j < 15; j++) {
-        nameBuilder.append((char) (buffer.get() & 0xFF));
+      int nodeStatus = buffer.getChar();
+      if (nodeStatus != 0x20 && nodeStatus != 0x21) {
+        throw new BrowsingException("Received negative response for the broadcast");
       }
 
-      final String serverName = nameBuilder.toString().trim();
-      int type = buffer.get();
+      skipBytes(buffer, 2);
+      skipBytes(buffer, 4);
+      skipBytes(buffer, 2);
+
+      int addressListEntryCount = buffer.get();
 
-      if (type == FILE_SERVER_NODE_TYPE) {
-        servers.add(serverName);
-      }
+      for (int i = 0; i < addressListEntryCount; i++) {
+        byte[] nameArray = new byte[SERVER_NAME_LENGTH];
+        buffer.get(nameArray, 0, SERVER_NAME_LENGTH);
 
-      skipBytes(buffer, 2);
+        final String serverName = new String(nameArray, Charset.forName(SERVER_NAME_CHARSET));
+        final int type = buffer.get();
+
+        if (type == FILE_SERVER_NODE_TYPE) {
+          servers.add(serverName);
+        }
+
+        skipBytes(buffer, 2);
+      }
+    } catch (BufferUnderflowException e) {
+      Log.e(TAG, "Malformed incoming packet");
```

Comment:
Mismatch of transId didn't really seem to be an error to me. Just sth that need to be filtered out. 

I didn't want to let BufferUnderflowException propagate and catch it outside like I do BrowsingException because ByteBuffer is parser's implementation detail. I didn't want to rethrow it as a BrowsingException to avoid creating a new object. But I'm not sure now if it should be done at the cost of clarity/consistency.